src/org/sosy_lab/cpachecker/cfa/parser/llvm/CFABuilder.java

```diff
@@ -1421,6 +1447,28 @@ private CSimpleDeclaration getAssignedVarDeclaration(
     return variableDeclarations.get(itemId);
   }
 
+  private CExpression castToExpectedType(CExpression expression, final CType pExpectedType,
+                                         final FileLocation location) {
+    CType expressionType = expression.getExpressionType();
+
+    if (expressionType.canBeAssignedFrom(pExpectedType)) {
+      return expression;
+
+    } else if (pointerOf(pExpectedType, expressionType)) {
+      CType typePointingTo = ((CPointerType) pExpectedType).getType().getCanonicalType();
+      if (expressionType.canBeAssignedFrom(typePointingTo)
+          || expressionType.equals(typePointingTo)) {
+        return getReference(location, expression);
+      } else {
+        throw new AssertionError("Unhandled type structure");
+      }
+    } else if (expressionType instanceof CPointerType) {
+      return getDereference(location, expression);
+    } else {
+      throw new AssertionError("Unhandled types structure");
```

Comment:
I would add the expressionType to the assert message, so that users know which type created the problem

Reply:
Sounds reasonable.